File: backend/common/money.py
from decimal import Decimal, ROUND_HALF_UP


CENT = Decimal("0.01")


def convert_money(amount: Decimal, source: str, target: str, usd_to_thb: Decimal) -> Decimal:
    """Convert THB/USD without binary floating-point loss."""
    if source not in {"THB", "USD"} or target not in {"THB", "USD"}:
        raise ValueError("Unsupported currency")
    if usd_to_thb <= 0:
        raise ValueError("Exchange rate must be positive")
    if source == target:
        converted = amount
    elif source == "USD":
        converted = amount * usd_to_thb
    else:
        converted = amount / usd_to_thb
    return converted.quantize(CENT, rounding=ROUND_HALF_UP)

File: frontend/src/lib/money.ts
import type { AppLang, Currency } from "../types";


export function formatMoney(amount: number, currency: Currency, locale: AppLang): string {
  const formatted = new Intl.NumberFormat(locale === "th" ? "th-TH" : "en-US", {
    minimumFractionDigits: 2,
    maximumFractionDigits: 2,
  }).format(amount);
  return `${currency === "THB" ? "฿" : "$"}${formatted}`;
}


export function convertAmount(
  amount: number,
  source: Currency,
  target: Currency,
  usdToThb: number,
): number {
  if (usdToThb <= 0) throw new Error("Exchange rate must be positive");
  if (source === target) return amount;
  const converted = source === "USD" ? amount * usdToThb : amount / usdToThb;
  return Math.round((converted + Number.EPSILON) * 100) / 100;
}

